nx_cugraph/algorithms/bipartite/matrix.py

```diff
@@ -0,0 +1,127 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import cupy as cp
+import networkx as nx
+import numpy as np
+
+from nx_cugraph.convert import _to_graph
+from nx_cugraph.generators._utils import _create_using_class
+from nx_cugraph.utils import index_dtype, networkx_algorithm
+
+__all__ = ["biadjacency_matrix", "from_biadjacency_matrix"]
+
+
+@networkx_algorithm(version_added="25.06")
+def biadjacency_matrix(
+    G, row_order, column_order=None, dtype=None, weight="weight", format="csr"
+):
+    import scipy as sp
+
+    G = _to_graph(G, weight, 1, dtype)
+
+    nrows = len(row_order)
+    if nrows == 0:
+        raise nx.NetworkXError("row_order is empty list")
+    if len(row_order) != len(set(row_order)):
+        raise nx.NetworkXError("Ambiguous ordering: `row_order` contained duplicates.")
+    if column_order is None:
+        column_order = list(set(G) - set(row_order))
+    ncols = len(column_order)
+    if len(column_order) != len(set(column_order)):
+        raise nx.NetworkXError(
+            "Ambiguous ordering: `column_order` contained duplicates."
+        )
+
+    if G.src_indices.size == 0:
+        src_indices = dst_indices = edge_array = ()
+    else:
+        row_ids = G._nodekeys_to_nodearray(row_order)
+        col_ids = G._nodekeys_to_nodearray(column_order)
+
+        # Using `mapper` like this is a useful trick that may not be obvious.
+        # This is also done in `to_scipy_sparse_array`.
+        mapper = cp.empty(G._N, dtype=index_dtype)
+        mapper[:] = -1  # Indicate nodes to exclude
+        mapper[row_ids] = cp.arange(row_ids.size, dtype=index_dtype)
+        src_indices = mapper[G.src_indices]
+
+        mapper[:] = -1  # Indicate nodes to exclude
+        mapper[col_ids] = cp.arange(col_ids.size, dtype=index_dtype)
+        dst_indices = mapper[G.dst_indices]
+        mask = (src_indices != -1) & (dst_indices != -1)
+        src_indices = src_indices[mask]
+        if src_indices.size == 0:
+            src_indices = dst_indices = edge_array = ()
+        else:
+            dst_indices = dst_indices[mask]
+
+            src_indices = cp.asnumpy(src_indices)
+            dst_indices = cp.asnumpy(dst_indices)
+            if weight in G.edge_values:
+                edge_array = G.edge_values[weight]
+                if weight in G.edge_masks:
+                    edge_array = cp.where(G.edge_masks[weight], edge_array, 1)
+                edge_array = edge_array[mask]
+                edge_array = cp.asnumpy(edge_array)
+            else:
+                edge_array = np.repeat(1, src_indices.size)
+
+    # PERF: convert to desired sparse format on GPU before copying to CPU
+    A = sp.sparse.coo_array(
```

Comment:
I don't remember the specifics but I thought we/the community decided that backends returning duck-types is acceptable, and if so, can we skip the copy to CPU entirely and return a cupy COO, or are cupy objects too different? Seems like if we can get away with returning cupy objects then there could be big benefits to keeping data on GPU as long as possible.

Reply:
Just like cupy arrays are not duck-compatible as numpy arrays, neither are cupyx sparse arrays compatible with scipy sparse arrays.

We've discussed adding an extra keyword argument to keep the result on device. We could also pursue the `cudf.pandas`-like behavior of a proxy object that can be on host, device, or both.

We can also open an issue about "from scipy array" functions _accepting_ cupyx sparse arrays.